lib/Runtime/Library/AtomicsObject.cpp

```diff
@@ -0,0 +1,270 @@
+            }
+        }
+
+        Assert(TypedArrayBase::Is(typedArray));
+        
+        if (accessIndex < 0 || accessIndex >= (int32)TypedArrayBase::FromVar(typedArray)->GetLength())
+        {
+            JavascriptError::ThrowRangeError(scriptContext, JSERR_InvalidTypedArrayIndex);
+        }
+
+        return (uint32)accessIndex;
+    }
+
+    TypedArrayBase * AtomicsObject::ValidateAndGetTypedArray(Var typedArray, Var index, __out uint32 *accessIndex, ScriptContext *scriptContext, bool onlyInt32)
+    {
+        ValidateSharedIntegerTypedArray(typedArray, scriptContext, onlyInt32);
+        uint32 i = ValidateAtomicAccess(typedArray, index, scriptContext);
+        if (accessIndex != nullptr)
+        {
+            *accessIndex = i;
+        }
+        Assert(TypedArrayBase::Is(typedArray));
+        return TypedArrayBase::FromVar(typedArray);
+    }
+
+    Var AtomicsObject::EntryAdd(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.add");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedAdd(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntryAnd(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.and");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedAnd(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntryCompareExchange(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(4, "Atomics.compareExchange");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+
+        Var comparand = args[3];
+        Var replacementValue = args[4];
+
+        return typedArrayBase->TypedCompareExchange(accessIndex, comparand, replacementValue);
+    }
+
+    Var AtomicsObject::EntryExchange(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.exchange");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedExchange(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntryIsLockFree(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(1, "Atomics.isLockFree");
+        uint32 size = JavascriptConversion::ToUInt32(args[1], scriptContext);
+        return (size == 1 || size == 2 || size == 4) ? scriptContext->GetLibrary()->GetTrue() : scriptContext->GetLibrary()->GetFalse();
```

Comment:
It is somewhat explained in the spec - but will add comment over here.

---
In reply to: [77404558](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77404558) [](ancestors = 77404558)